x/expiration/types/expiration.go

```diff
@@ -0,0 +1,100 @@
+package types
+
+import (
+	"fmt"
+	"regexp"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/cosmos/cosmos-sdk/codec/types"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+)
+
+func NewExpiration(
+	moduleAssetID string,
+	owner string,
+	time time.Time,
+	deposit sdk.Coins,
+	message types.Any,
+) *Expiration {
+	return &Expiration{
+		ModuleAssetId: moduleAssetID,
+		Owner:         owner,
+		Time:          time,
+		Deposit:       deposit,
+		Message:       message,
+	}
+}
+
+// ValidateBasic basic format checking of the data
+func (e *Expiration) ValidateBasic() error {
+	if strings.TrimSpace(e.ModuleAssetId) == "" {
+		return ErrEmptyModuleAssetID
+	}
+	if err := validateAddress(e.ModuleAssetId); err != nil {
+		return err
+	}
+	if strings.TrimSpace(e.Owner) == "" {
+		return ErrEmptyOwnerAddress
+	}
+	if e.Time.IsZero() {
+		return ErrExpirationTime
+	}
+	if e.Deposit.IsZero() || e.Deposit.IsAnyNil() || !e.Deposit.IsValid() {
+		return ErrInvalidDeposit
+	}
+	if err := e.validateMessage(); err != nil {
+		return ErrInvalidMessage
+	}
+	return nil
+}
+
+// validateMessage validates `expiration.Message` conforms to `sdk.Msg`
+// and is registered and whitelisted with the InterfaceRegistry
+func (e *Expiration) validateMessage() error {
+	// this may occur in message decoding
+	if e.Message.Value == nil {
+		return fmt.Errorf("expecting non nil value to validate an Any message")
+	}
+	if e.Message.TypeUrl == "" {
```

Comment:
Do we need to validate that this msg is of a certain TypeUrl?

Reply:
Yes, but there's no mechanism in place at the moment that allows us to do that. (ref: https://github.com/provenance-io/provenance/pull/1115#discussion_r987141369)